### PR TITLE
Add 'mkVersionIntervals' for creating a VersionIntervals from a list

### PR DESCRIPTION
**Pull request by dan-t**
I tried updating my program cabal-bounds[1] to the newest cabal version, but wasn't able anymore to create a 'VersionIntervals' from a '[VersionInterval]'.

This patch adds back the 'mkVersionIntervals' function.

[1] https://github.com/dan-t/cabal-bounds

---
* [ X] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).

### CODE REVIEW

**Comment by ulysses4ever** [comment]
Hello! The function you're using is there, just in another module:

https://github.com/dan-t/cabal/blob/e71f6f263aa4d7ce7a145eb5ac417f2f580f2288/Cabal-syntax/src/Distribution/Types/VersionInterval/Legacy.hs#L179-L186

(You can read the module's header on a brief history of transitions.)

Would it be enough for your purposes?

**Comment by dan-t** [comment]
Hi Artem,

> The function you're using is there, just in another module:
> 
> https://github.com/dan-t/cabal/blob/e71f6f263aa4d7ce7a145eb5ac417f2f580f2288/Cabal-syntax/src/Distribution/Types/
> VersionInterval/Legacy.hs#L179-L186
> 
> (You can read the module's header on a brief history of transitions.)
> 
> Would it be enough for your purposes?

no, because the types 'VersionInterval' and 'VersionIntervals' in 'Distribution.Types.VersionInterval.Legacy'
and 'Distribution.Types.VersionInterval' aren't compatible.

Greetings,
Daniel

**Comment by ulysses4ever** [comment]
All right. You say the PR "adds back" the function. Can you print when it was removed? Then we could read the rational for its removal (if its there) and decide whether it's a good idea to add it back.

**Comment by dan-t** [comment]
I don‘t know when the function was removed. 

Without this function the cabal library can‘t be used to change the version intervals of dependencies. So my program cabal-bounds[1] can‘t work with newer versions of the cabal library.

[1] https://github.com/dan-t/cabal-bounds

**Comment by ffaf1** [comment]
`mkVersionIntervals` is still there, just moved in `Legacy` (function is exported, module is exposed):

https://github.com/haskell/cabal/blob/94c86abdf1ea4d9e79a28b14ca003f381eadf974/Cabal-syntax/src/Distribution/Types/VersionInterval/Legacy.hs#L178-L186

1347791c7bf70b1bf54edc30acbea93bad5d46d0 / #7034

Does this solve your problem, @dan-t?

**Comment by ulysses4ever** [comment]
@ffaf1 did you see discussion above?

**Comment by ffaf1** [comment]
I haven't and I should have! At least now we have a commit/PR.

**Comment by chreekat** [comment]
Looks to have been moved to the Legacy module in 1347791c7bf70b1bf54edc30acbea93bad5d46d0. Maybe @phadej can say more.

I think I can see a way to convert `[VersionInterval]` to `VersionRange` which then could be used to get a `VersionIntervals`, but I am not in any way qualified to comment on that idea.

**Comment by ulysses4ever** [comment]
All right, I don't see an excuse for us to hold this any longer. This is simple enough function,and I don't see a reason to reject it. So, I'm willing to approve modulo changelog.

@dan-t this is a API change, so it needs a changelog file (see https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).

**Review by andreasabel — APPROVED** [review]
As can be seen from the discussion in #7916, my opinion is that the non-legacy API to version intervals is broken, so, yes, I approve attempts to fix it.

**Comment by dan-t** [comment]
Ok, I've added the changelog entry.

File: changelog.d/pr-9034
Comment (ulysses4ever) [inline changelog.d/pr-9034:2]:
```suggestion
packages: Cabal-syntax
```

File: changelog.d/pr-9034
Comment (ulysses4ever) [inline changelog.d/pr-9034:7]:
```suggestion
If external tools want to change the version intervals of dependencies, they
```

File: changelog.d/pr-9034
Comment (ulysses4ever) [inline changelog.d/pr-9034:8]:
```suggestion
need to be able to create a `VersionRange` from a `[VersionInterval]` list. Adding
```

File: changelog.d/pr-9034
Comment (ulysses4ever) [inline changelog.d/pr-9034:9]:
```suggestion
the function `mkVersionIntervals` makes this possible again.
```

File: changelog.d/pr-9034
Comment (ulysses4ever) [inline changelog.d/pr-9034:1]:
```suggestion
synopsis: Add `mkVersionIntervals` for creating a `VersionIntervals` from a list
```

File: changelog.d/pr-9034
Comment (ulysses4ever) [inline changelog.d/pr-9034:8]:
```suggestion
need to be able to create a `VersionRange` from a `[VersionInterval]` list. Adding
```

**Review by ulysses4ever — APPROVED** [review]
Looks good, thanks! Formating suggestions pending.

**Comment by ulysses4ever** [comment]
@dan-t there's a bunch of formating suggestions. You accepted one. I'm just checking that your didn't miss others.

**Comment by dan-t** [comment]
﻿
@dan-t there's a bunch of formating suggestions. You accepted one. I'm just checking that your didn't miss others.Oh sorry, I haven‘t been aware about it. So I haven‘t to do now anything, right?

**Comment by ulysses4ever** [comment]
@dan-t no worries! GitHub UI around suggestions is very confusing. I just applied the suggestions but in general it'd be the author's job. 

All good now. There's will be a several-day delay before merge.

File: Cabal-syntax/src/Distribution/Types/VersionInterval.hs
Comment (phadej) [inline Cabal-syntax/src/Distribution/Types/VersionInterval.hs:75]: FWIW, this is different signature than in the legacy module. That one doesn't have `Maybe`.

Also if the goal is to make `[VersionInterval] -> VersionRange` function, why just not export that, i.e. a variant of `fromVersionIntervals` with that signature. It works even if invariants are violated (e.g. if intervals overlap).

**Comment by dan-t** [comment]
A function that silently ignores violations of invariants isn‘t something I would consider good API design.

Why not having a function that just creates a VersionRange? I wanted to apply the most minimal change to the current code base, just adding back the function that was removed.

**Comment by phadej** [comment]
>  I wanted to apply the most minimal change to the current code base, just adding back the function that was removed.

You added **different function**, you added a function with `Maybe`.  The `Legacy` function is

```haskell

-- | Directly construct a 'VersionIntervals' from a list of intervals.
--
mkVersionIntervals :: [VersionInterval] -> VersionIntervals
mkVersionIntervals intervals
    | invariant (VersionIntervals intervals) = VersionIntervals intervals
    | otherwise
        = checkInvariant
        . foldl' (flip insertInterval) (VersionIntervals [])
        . filter validInterval
        $ intervals
```

which reintroduces the invariants if needed.

> A function that silently ignores violations of invariants isn‘t something I would consider good API design.

I didn't propose such. I tried to say that `[VersionInterval] -> VersionRange` can be written (and will work correctly) even if `[VersionInterval]`s don't satisfy `VersionIntervals` invariants (like not overlapping). In that case you may create `(>= 1 && <3) || (>= 2 && <4)`-like `VersionRange`, but it's a valid one, just not normalised.

**Comment by ulysses4ever** [comment]
@phadej 

> You added different function

You seem to suggest that it's a bad thing. Why? Do we have to follow the Legacy module and not introduce anything new?

**Comment by phadej** [comment]
@ulysses4ever because `Maybe` maybe in the way. Probably user amend `[VersionInterval]` in some way, and if they (accidentally) break the invariant of `VersionIntervals` (which they probably don't care about), they will get `Nothing`. Not very useful.

**Comment by dan-t** [comment]
At the end you guys have to decide which function you would like to
have. I don't care that much how the function looks and behaves, as long
as I can create a `VersionIntervals` from a `[VersionInterval]`, or even
directly a `VersionRange` from a `[VersionInterval]`.

**Comment by ulysses4ever** [comment]
> because Maybe maybe in the way.

Or maybe not. If you program in Haskell, you probably know how to deal with it.

**Comment by ulysses4ever** [comment]
@Mergify rebase

**Pull request merged**
